core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -1281,6 +1284,81 @@ public MaterializedViewFreshness getMaterializedViewFreshness(Session session, Q
         return metadata.applyTableScanRedirect(connectorSession, tableHandle.getConnectorHandle());
     }
 
+    @Override
+    public QualifiedObjectName getRedirectedTableName(Session session, QualifiedObjectName originalTableName)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(originalTableName, "originalTableName is null");
+
+        QualifiedObjectName tableName = originalTableName;
+        Set<QualifiedObjectName> visitedTableNames = new LinkedHashSet<>();
+        for (int count = 0; count < MAX_TABLE_REDIRECTIONS; count++) {
+            // Check for loop in redirection
+            if (!visitedTableNames.add(tableName)) {
+                throw new TrinoException(TABLE_REDIRECTION_ERROR,
+                        format("Table redirections form a loop: %s",
+                                getRedirectionChain(ImmutableList.<QualifiedObjectName>builder()
+                                        .addAll(visitedTableNames)
+                                        .add(tableName)
+                                        .build())));
+            }
+
+            Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, tableName.getCatalogName());
+            if (catalog.isPresent()) {
+                CatalogMetadata catalogMetadata = catalog.get();
+                CatalogName catalogName = catalogMetadata.getConnectorId(session, tableName);
+                ConnectorMetadata metadata = catalogMetadata.getMetadataFor(catalogName);
+
+                Optional<QualifiedObjectName> redirectedTableName = metadata.redirectTable(session.toConnectorSession(catalogName), tableName.asSchemaTableName())
+                        .map(name -> convertFromSchemaTableName(name.getCatalogName()).apply(name.getSchemaTableName()));
+
+                if (redirectedTableName.isEmpty()) {
+                    return tableName;
+                }
+
+                tableName = redirectedTableName.get();
+            }
+            else {
+                // Stop redirection
+                return tableName;
+            }
+        }
+        throw new TrinoException(TABLE_REDIRECTION_ERROR, format("Table redirected too many times (%d): %s", MAX_TABLE_REDIRECTIONS, visitedTableNames));
+    }
+
+    private static String getRedirectionChain(Collection<QualifiedObjectName> names)
+    {
+        return names.stream()
+                .map(QualifiedObjectName::toString)
+                .collect(Collectors.joining(" -> "));
+    }
+
+    @Override
+    public Optional<TableHandle> getRedirectedTableHandle(Session session, QualifiedObjectName tableName)
+    {
+        QualifiedObjectName targetTableName = getRedirectedTableName(session, tableName);
+
+        if (!targetTableName.equals(tableName)) {
+            Optional<TableHandle> tableHandle = getTableHandle(session, targetTableName);
+
+            if (tableHandle.isEmpty()) {
+                // Redirected table must exist
```

Comment:
If I'm understanding the scenario, this means a connector returned a redirection, then the target did not exist. Why would we want to fallback?

In the case of the Hive connector redirecting to Iceberg, the subsequent read operation would fail (maybe in a confusing way). Though this might be a bad example, since they share the same metastore and thus this should not occur.

Reply:
I was wondering whether in case of problems in finding the target table (could be a bad redirection config or target connector metastore is unavailable temporarily), a silent fallback (with warning) is useful to avoid disrupting normal operations. 
A contrived example would be that the target table is in memory connector, then the coordinator was restarted, now the target table doesn't exist. The source connector probably can't easily detect status of target table.
However, it's also useful to get an explicit error in case one doesn't want to fallback to source at all.
Would we have a catalog session property in each connector which implements this redirection to disable the redirection if needed ?

Reply:
IMO it'd be better to fail-fast in this case with a clear error, since the target table is "externally" provided by the connector, so that the engine code doesn't need to be aware of throwing redirection-related errors - whenever it needs a handle. 

Having a catalog property switch seems like a good idea.